### PR TITLE
Differentiate prelude constructor and custom type constructors

### LINKED ISSUE #1112: JS: Bool and Nil should only be special cased if we know that they are the prelude ones
Searach for the URL of this issue to find the problem in the codebase.

We will need to add `Type` to the `Pattern::Constructor` variant and check it when rendering patterns in JS.

### PR DESCRIPTION
**Pull request by shunjilin**
closes #1112 

Add `type_` field to `Pattern` constructor to allow js to differentiate between prelude constructor and custom type constructor

### CODE REVIEW

File: src/parse.rs [hunk 1]
```diff
@@ -1040,6 +1040,7 @@ where
             name,
             with_spread,
             constructor: (),
+            type_: (),
```
Comment (shunjilin) [inline src/parse.rs:1043]: Looking at the signature I assumed this is untyped, not sure if putting `()` here is correct

**Review by lpil — COMMENTED** [review]
Looks great!

As a bit of a nit-pick can we keep difference match clauses for true, false, and nil that isn't with the other record clause? I find it clearer if it's split out.

We could use something like this:

```rust
            Pattern::Constructor {
                constructor: PatternConstructor::Record { type_, .. },
                ..
            } if type_.is_nil() => {
                self.booly_check(false);
                Ok(())
            }
```

**Comment by shunjilin** [comment]
> 
> 
> Looks great!
> 
> As a bit of a nit-pick can we keep difference match clauses for true, false, and nil that isn't with the other record clause? I find it clearer if it's split out.
> 
> We could use something like this:
> 
> ```rust
>             Pattern::Constructor {
>                 constructor: PatternConstructor::Record { type_, .. },
>                 ..
>             } if type_.is_nil() => {
>                 self.booly_check(false);
>                 Ok(())
>             }
> ```

Thank you! Fixed, and looks cleaner indeed

> 
> 
> Looks great!
> 
> As a bit of a nit-pick can we keep difference match clauses for true, false, and nil that isn't with the other record clause? I find it clearer if it's split out.
> 
> We could use something like this:
> 
> ```rust
>             Pattern::Constructor {
>                 constructor: PatternConstructor::Record { type_, .. },
>                 ..
>             } if type_.is_nil() => {
>                 self.booly_check(false);
>                 Ok(())
>             }
> ```

Thank you! Fixed, and looks cleaner indeed

**Review by lpil — APPROVED** [review]
Wonderful work! Thank you! ✨

**Comment by lpil** [comment]
CI is broken because of a mistake I made, can ignore that

**Pull request merged**